runtime/include/QuantumDevice.hpp

```diff
@@ -148,17 +148,53 @@ struct QuantumDevice {
      * @param inverse Indicates whether to use inverse of gate
      */
     virtual void NamedOperation(const std::string &name, const std::vector<double> &params,
-                                const std::vector<QubitIdType> &wires, bool inverse) = 0;
+                                const std::vector<QubitIdType> &wires, bool inverse)
+    {
+        NamedOperation(name, params, wires, inverse, {}, {});
+    }
+
+    /**
+     * @brief Apply a single gate to the state vector of a device with its name if this is
+     * supported.
+     *
+     * @param name The name of the gate to apply
+     * @param params Optional parameter list for parametric gates
+     * @param wires Wires to apply gate to
+     * @param inverse Indicates whether to use inverse of gate
+     * @param controlled_wires Controlled wires applied to the operation
+     * @param controlled_values Controlled values applied to the operation
+     */
+    virtual void NamedOperation(const std::string &name, const std::vector<double> &params,
+                                const std::vector<QubitIdType> &wires, bool inverse,
+                                const std::vector<QubitIdType> &controlled_wires,
+                                const std::vector<bool> &controlled_values) = 0;
+
+    /**
+     * @brief Apply a given matrix directly to the state vector of a device.
+     *
+     * @param matrix The matrix of data in row-major format
+     * @param wires Wires to apply gate to
+     * @param inverse Indicates whether to use inverse of gate
+     */
+    virtual void MatrixOperation(const std::vector<std::complex<double>> &matrix,
+                                 const std::vector<QubitIdType> &wires, bool inverse)
+    {
+        MatrixOperation(matrix, wires, inverse, {}, {});
+    }
```

Comment:
Let's remove this one too if we end up removing the non-controlled `NamedOperation`.

Reply:
Done, thanks!